File: python/bonsai-reference/src/bonsai_reference/stats.py
"""Independent integer statistical reference for BK-13.

Paired mean differences, deterministic bootstrap intervals, Holm adjustment,
and undeclared-exclusion detection. No post-hoc seed or metric selection.
"""

from __future__ import annotations

from dataclasses import dataclass, replace
from typing import Final

LCG_A: Final[int] = 1_664_525
LCG_C: Final[int] = 1_013_904_223
LCG_MASK: Final[int] = 0xFFFF_FFFF


class StatError(ValueError):
    """Stable statistical-plan failure."""

    def __init__(self, code: str) -> None:
        self.code = code
        super().__init__(code)


@dataclass(frozen=True, slots=True)
class SeedPair:
    pair_id: str
    outcome_a: int | None
    outcome_b: int | None
    failed_a: bool
    failed_b: bool


@dataclass(frozen=True, slots=True)
class OutcomeFamily:
    outcome_id: str
    pairs: tuple[SeedPair, ...]


@dataclass(frozen=True, slots=True)
class StatisticalPlan:
    preregistered_outcomes: tuple[str, ...]
    declared_pairs: tuple[str, ...]
    declared_exclusions: tuple[str, ...]
    bootstrap_resamples: int
    bootstrap_seed: int
    alpha_numerator: int
    alpha_denominator: int


@dataclass(frozen=True, slots=True)
class OutcomeRow:
    outcome_id: str
    complete_pairs: int
    missing_runs: int
    failed_runs: int
    excluded_runs: int
    effect_numerator: int | None
    effect_denominator: int | None
    ci_low_numerator: int | None
    ci_high_numerator: int | None
    ci_denominator: int | None
    p_numerator: int | None
    p_denominator: int | None
    holm_rejected: bool
    sensitivity_min_numerator: int | None
    sensitivity_max_numerator: int | None
    sensitivity_denominator: int | None
    detail_code: str | None


def aggregate(plan: StatisticalPlan, families: tuple[OutcomeFamily, ...]) -> tuple[OutcomeRow, ...]:
    """Aggregate preregistered paired outcomes and reject undeclared exclusions."""
    _validate_plan(plan)
    declared = set(plan.declared_pairs)
    excluded = set(plan.declared_exclusions)
    preregistered = set(plan.preregistered_outcomes)
    if declared & excluded:
        raise StatError("STAT_PLAN_INVALID")
    rows: list[OutcomeRow] = []
    p_values: list[tuple[str, int, int]] = []
    for family in families:
        if family.outcome_id not in preregistered:
            raise StatError("STAT_UNDECLARED_METRIC")
        diffs, missing, failed, excluded_runs = _complete_diffs(family.pairs, declared, excluded)
        if not diffs:
            rows.append(
                OutcomeRow(
                    outcome_id=family.outcome_id,
                    complete_pairs=0,
                    missing_runs=missing,
                    failed_runs=failed,
                    excluded_runs=excluded_runs,
                    effect_numerator=None,
                    effect_denominator=None,
                    ci_low_numerator=None,
                    ci_high_numerator=None,
                    ci_denominator=None,
                    p_numerator=None,
                    p_denominator=None,
                    holm_rejected=False,
                    sensitivity_min_numerator=None,
                    sensitivity_max_numerator=None,
                    sensitivity_denominator=None,
                    detail_code="STAT_PAIRS_UNAVAILABLE",
                )
            )
            continue
        effect = sum(diffs)
        low, high = _bootstrap_ci(diffs, plan.bootstrap_resamples, plan.bootstrap_seed)
        p_num, p_den = _bootstrap_p(diffs, plan.bootstrap_resamples, plan.bootstrap_seed)
        sens_min, sens_max, sens_den = _sensitivity(diffs)
        rows.append(
            OutcomeRow(
                outcome_id=family.outcome_id,
                complete_pairs=len(diffs),
                missing_runs=missing,
                failed_runs=failed,
                excluded_runs=excluded_runs,
                effect_numerator=effect,
                effect_denominator=len(diffs),
                ci_low_numerator=low,
                ci_high_numerator=high,
                ci_denominator=len(diffs),
                p_numerator=p_num,
                p_denominator=p_den,
                holm_rejected=False,
                sensitivity_min_numerator=sens_min,
                sensitivity_max_numerator=sens_max,
                sensitivity_denominator=sens_den,
                detail_code=None,
            )
        )
        p_values.append((family.outcome_id, p_num, p_den))
    rejected = _holm(p_values, plan.alpha_numerator, plan.alpha_denominator)
    return tuple(
        row if row.outcome_id not in rejected else replace(row, holm_rejected=True)
        for row in sorted(rows, key=lambda item: item.outcome_id)
    )


def _validate_plan(plan: StatisticalPlan) -> None:
    if (
        not plan.preregistered_outcomes
        or not plan.declared_pairs
        or plan.bootstrap_resamples < 2
        or plan.alpha_numerator <= 0
        or plan.alpha_denominator == 0
        or len(set(plan.preregistered_outcomes)) != len(plan.preregistered_outcomes)
        or len(set(plan.declared_pairs)) != len(plan.declared_pairs)
        or len(set(plan.declared_exclusions)) != len(plan.declared_exclusions)
    ):
        raise StatError("STAT_PLAN_INVALID")


def _complete_diffs(
    pairs: tuple[SeedPair, ...],
    declared: set[str],
    excluded: set[str],
) -> tuple[list[int], int, int, int]:
    diffs: list[int] = []
    missing = 0
    failed = 0
    excluded_runs = 0
    for pair in sorted(pairs, key=lambda item: item.pair_id):
        if pair.pair_id not in declared and pair.pair_id not in excluded:
            raise StatError("STAT_UNDECLARED_EXCLUSION")
        if pair.pair_id in excluded:
            excluded_runs += 1
            continue
        if pair.failed_a or pair.failed_b:
            failed += 1
            continue
        if pair.outcome_a is None or pair.outcome_b is None:
            missing += 1
            continue
        diffs.append(pair.outcome_a - pair.outcome_b)
    return diffs, missing, failed, excluded_runs


def _lcg(state: int) -> int:
    return (state * LCG_A + LCG_C) & LCG_MASK


def _bootstrap_sums(diffs: list[int], resamples: int, seed: int) -> list[int]:
    state = seed & LCG_MASK
    n = len(diffs)
    sums: list[int] = []
    for _ in range(resamples):
        total = 0
        for _ in range(n):
            state = _lcg(state)
            total += diffs[state % n]
        sums.append(total)
    return sums


def _bootstrap_ci(diffs: list[int], resamples: int, seed: int) -> tuple[int, int]:
    sums = sorted(_bootstrap_sums(diffs, resamples, seed))
    low_index = (25 * (resamples - 1)) // 1000
    high_index = (975 * (resamples - 1)) // 1000
    return sums[low_index], sums[high_index]


def _bootstrap_p(diffs: list[int], resamples: int, seed: int) -> tuple[int, int]:
    observed = sum(diffs)
    count = 0
    for total in _bootstrap_sums(diffs, resamples, seed):
        if observed == 0 or (observed > 0 and total <= 0) or (observed < 0 and total >= 0):
            count += 1
    return count + 1, resamples + 1


def _sensitivity(diffs: list[int]) -> tuple[int | None, int | None, int | None]:
    if len(diffs) < 2:
        return None, None, None
    leave_one = [sum(diffs) - value for value in diffs]
    return min(leave_one), max(leave_one), len(diffs) - 1


def _holm(p_values: list[tuple[str, int, int]], alpha_num: int, alpha_den: int) -> set[str]:
    ordered = sorted(p_values, key=lambda item: (item[1] * 10_000 // item[2], item[0]))
    rejected: set[str] = set()
    count = len(ordered)
    for index, (outcome_id, p_num, p_den) in enumerate(ordered):
        remaining = count - index
        if p_num * alpha_den * remaining <= p_den * alpha_num:
            rejected.add(outcome_id)
        else:
            break
    return rejected


def diagnostic_plan() -> StatisticalPlan:
    return StatisticalPlan(
        preregistered_outcomes=("latency", "reward"),
        declared_pairs=("p1", "p2", "p3", "p4"),
        declared_exclusions=("p_skip",),
        bootstrap_resamples=20,
        bootstrap_seed=1,
        alpha_numerator=5,
        alpha_denominator=100,
    )


def diagnostic_families() -> tuple[OutcomeFamily, ...]:
    reward = OutcomeFamily(
        outcome_id="reward",
        pairs=(
            SeedPair("p1", 10, 6, False, False),
            SeedPair("p2", 8, 7, False, False),
            SeedPair("p3", 12, 5, False, False),
            SeedPair("p4", 9, None, False, True),
            SeedPair("p_skip", 100, 0, False, False),
        ),
    )
    latency = OutcomeFamily(
        outcome_id="latency",
        pairs=(
            SeedPair("p1", 3, 4, False, False),
            SeedPair("p2", 2, 5, False, False),
            SeedPair("p3", 1, 6, False, False),
            SeedPair("p4", None, 2, True, False),
            SeedPair("p_skip", 0, 50, False, False),
        ),
    )
    return (latency, reward)

File: crates/bonsai-metrics/src/stats.rs
//! Paired statistical aggregation with Holm adjustment and undeclared-exclusion detection.

use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;
use std::error::Error;
use std::fmt;

const LCG_A: u64 = 1_664_525;
const LCG_C: u64 = 1_013_904_223;
const LCG_MASK: u64 = 0xFFFF_FFFF;

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct SeedPair {
    pub pair_id: String,
    pub outcome_a: Option<i64>,
    pub outcome_b: Option<i64>,
    pub failed_a: bool,
    pub failed_b: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct OutcomeFamily {
    pub outcome_id: String,
    pub pairs: Vec<SeedPair>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct StatisticalPlan {
    pub preregistered_outcomes: Vec<String>,
    pub declared_pairs: Vec<String>,
    pub declared_exclusions: Vec<String>,
    pub bootstrap_resamples: u64,
    pub bootstrap_seed: u64,
    pub alpha_numerator: i64,
    pub alpha_denominator: u64,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct OutcomeRow {
    pub outcome_id: String,
    pub complete_pairs: u64,
    pub missing_runs: u64,
    pub failed_runs: u64,
    pub excluded_runs: u64,
    pub effect_numerator: Option<i64>,
    pub effect_denominator: Option<u64>,
    pub ci_low_numerator: Option<i64>,
    pub ci_high_numerator: Option<i64>,
    pub ci_denominator: Option<u64>,
    pub p_numerator: Option<u64>,
    pub p_denominator: Option<u64>,
    pub holm_rejected: bool,
    pub sensitivity_min_numerator: Option<i64>,
    pub sensitivity_max_numerator: Option<i64>,
    pub sensitivity_denominator: Option<u64>,
    pub detail_code: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct StatisticalTable {
    pub schema: String,
    pub rows: Vec<OutcomeRow>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum StatError {
    Plan,
    UndeclaredExclusion,
    UndeclaredMetric,
}

impl fmt::Display for StatError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::Plan => "STAT_PLAN_INVALID",
            Self::UndeclaredExclusion => "STAT_UNDECLARED_EXCLUSION",
            Self::UndeclaredMetric => "STAT_UNDECLARED_METRIC",
        })
    }
}

impl Error for StatError {}

/// Aggregate preregistered paired outcomes.
///
/// Undeclared pair or metric omissions fail closed. Missing and failed runs
/// are reported and never silently dropped.
///
/// # Errors
///
/// Rejects an invalid plan, undeclared exclusions, or undeclared metrics.
pub fn aggregate(
    plan: &StatisticalPlan,
    families: &[OutcomeFamily],
) -> Result<StatisticalTable, StatError> {
    validate_plan(plan)?;
    let declared = plan.declared_pairs.iter().cloned().collect::<BTreeSet<_>>();
    let excluded = plan
        .declared_exclusions
        .iter()
        .cloned()
        .collect::<BTreeSet<_>>();
    if !declared.is_disjoint(&excluded) {
        return Err(StatError::Plan);
    }
    let preregistered = plan
        .preregistered_outcomes
        .iter()
        .cloned()
        .collect::<BTreeSet<_>>();
    let mut rows = Vec::new();
    let mut p_values = Vec::new();
    for family in families {
        if !preregistered.contains(&family.outcome_id) {
            return Err(StatError::UndeclaredMetric);
        }
        let completed = complete_diffs(&family.pairs, &declared, &excluded)?;
        if completed.diffs.is_empty() {
            rows.push(OutcomeRow {
                outcome_id: family.outcome_id.clone(),
                complete_pairs: 0,
                missing_runs: completed.missing,
                failed_runs: completed.failed,
                excluded_runs: completed.excluded,
                effect_numerator: None,
                effect_denominator: None,
                ci_low_numerator: None,
                ci_high_numerator: None,
                ci_denominator: None,
                p_numerator: None,
                p_denominator: None,
                holm_rejected: false,
                sensitivity_min_numerator: None,
                sensitivity_max_numerator: None,
                sensitivity_denominator: None,
                detail_code: Some("STAT_PAIRS_UNAVAILABLE".to_owned()),
            });
            continue;
        }
        let n = u64::try_from(completed.diffs.len()).unwrap_or(u64::MAX);
        let effect = completed.diffs.iter().sum::<i64>();
        let (ci_low, ci_high) = bootstrap_ci(
            &completed.diffs,
            plan.bootstrap_resamples,
            plan.bootstrap_seed,
        );
        let (p_num, p_den) = bootstrap_p(
            &completed.diffs,
            plan.bootstrap_resamples,
            plan.bootstrap_seed,
        );
        let (sens_min, sens_max, sens_den) = sensitivity(&completed.diffs);
        rows.push(OutcomeRow {
            outcome_id: family.outcome_id.clone(),
            complete_pairs: n,
            missing_runs: completed.missing,
            failed_runs: completed.failed,
            excluded_runs: completed.excluded,
            effect_numerator: Some(effect),
            effect_denominator: Some(n),
            ci_low_numerator: Some(ci_low),
            ci_high_numerator: Some(ci_high),
            ci_denominator: Some(n),
            p_numerator: Some(p_num),
            p_denominator: Some(p_den),
            holm_rejected: false,
            sensitivity_min_numerator: sens_min,
            sensitivity_max_numerator: sens_max,
            sensitivity_denominator: sens_den,
            detail_code: None,
        });
        p_values.push((family.outcome_id.clone(), p_num, p_den));
    }
    let rejected = holm(&p_values, plan.alpha_numerator, plan.alpha_denominator);
    for row in &mut rows {
        row.holm_rejected = rejected.contains(&row.outcome_id);
    }
    rows.sort_by(|left, right| left.outcome_id.cmp(&right.outcome_id));
    Ok(StatisticalTable {
        schema: "bonsai.statistical-table/v1".to_owned(),
        rows,
    })
}

struct Completed {
    diffs: Vec<i64>,
    missing: u64,
    failed: u64,
    excluded: u64,
}

fn validate_plan(plan: &StatisticalPlan) -> Result<(), StatError> {
    if plan.preregistered_outcomes.is_empty()
        || plan.declared_pairs.is_empty()
        || plan.bootstrap_resamples < 2
        || plan.alpha_numerator <= 0
        || plan.alpha_denominator == 0
        || plan
            .preregistered_outcomes
            .iter()
            .collect::<BTreeSet<_>>()
            .len()
            != plan.preregistered_outcomes.len()
        || plan.declared_pairs.iter().collect::<BTreeSet<_>>().len() != plan.declared_pairs.len()
        || plan
            .declared_exclusions
            .iter()
            .collect::<BTreeSet<_>>()
            .len()
            != plan.declared_exclusions.len()
    {
        return Err(StatError::Plan);
    }
    Ok(())
}

fn complete_diffs(
    pairs: &[SeedPair],
    declared: &BTreeSet<String>,
    excluded: &BTreeSet<String>,
) -> Result<Completed, StatError> {
    let mut ordered = pairs.to_vec();
    ordered.sort_by(|left, right| left.pair_id.cmp(&right.pair_id));
    let mut diffs = Vec::new();
    let mut missing = 0;
    let mut failed = 0;
    let mut excluded_runs = 0;
    for pair in ordered {
        if !declared.contains(&pair.pair_id) && !excluded.contains(&pair.pair_id) {
            return Err(StatError::UndeclaredExclusion);
        }
        if excluded.contains(&pair.pair_id) {
            excluded_runs += 1;
            continue;
        }
        if pair.failed_a || pair.failed_b {
            failed += 1;
            continue;
        }
        match (pair.outcome_a, pair.outcome_b) {
            (Some(left), Some(right)) => diffs.push(left.saturating_sub(right)),
            _ => missing += 1,
        }
    }
    Ok(Completed {
        diffs,
        missing,
        failed,
        excluded: excluded_runs,
    })
}

fn lcg(state: u64) -> u64 {
    state.wrapping_mul(LCG_A).wrapping_add(LCG_C) & LCG_MASK
}

fn bootstrap_sums(diffs: &[i64], resamples: u64, seed: u64) -> Vec<i64> {
    let mut state = seed & LCG_MASK;
    let n = diffs.len();
    let mut sums = Vec::new();
    for _ in 0..resamples {
        let mut total = 0_i64;
        for _ in 0..n {
            state = lcg(state);
            let index = usize::try_from(state % u64::try_from(n).unwrap_or(1)).unwrap_or(0);
            total = total.saturating_add(diffs[index]);
        }
        sums.push(total);
    }
    sums
}

fn bootstrap_ci(diffs: &[i64], resamples: u64, seed: u64) -> (i64, i64) {
    let mut sums = bootstrap_sums(diffs, resamples, seed);
    sums.sort_unstable();
    let last = resamples.saturating_sub(1);
    let low = usize::try_from((25 * last) / 1000).unwrap_or(0);
    let high = usize::try_from((975 * last) / 1000).unwrap_or(sums.len().saturating_sub(1));
    (sums[low], sums[high])
}

fn bootstrap_p(diffs: &[i64], resamples: u64, seed: u64) -> (u64, u64) {
    let observed = diffs.iter().sum::<i64>();
    let count = bootstrap_sums(diffs, resamples, seed)
        .into_iter()
        .filter(|total| {
            observed == 0 || (observed > 0 && *total <= 0) || (observed < 0 && *total >= 0)
        })
        .count();
    (
        u64::try_from(count).unwrap_or(u64::MAX).saturating_add(1),
        resamples.saturating_add(1),
    )
}

fn sensitivity(diffs: &[i64]) -> (Option<i64>, Option<i64>, Option<u64>) {
    if diffs.len() < 2 {
        return (None, None, None);
    }
    let total = diffs.iter().sum::<i64>();
    let leave = diffs
        .iter()
        .map(|value| total.saturating_sub(*value))
        .collect::<Vec<_>>();
    (
        leave.iter().copied().min(),
        leave.iter().copied().max(),
        Some(u64::try_from(diffs.len().saturating_sub(1)).unwrap_or(u64::MAX)),
    )
}

fn holm(p_values: &[(String, u64, u64)], alpha_num: i64, alpha_den: u64) -> BTreeSet<String> {
    let mut ordered = p_values.to_vec();
    ordered.sort_by(|left, right| {
        let left_key = left.1.saturating_mul(10_000) / left.2.max(1);
        let right_key = right.1.saturating_mul(10_000) / right.2.max(1);
        left_key.cmp(&right_key).then_with(|| left.0.cmp(&right.0))
    });
    let count = ordered.len();
    let mut rejected = BTreeSet::new();
    for (index, (outcome_id, p_num, p_den)) in ordered.into_iter().enumerate() {
        let remaining = u64::try_from(count.saturating_sub(index)).unwrap_or(1);
        let left = p_num.saturating_mul(alpha_den).saturating_mul(remaining);
        let right = p_den.saturating_mul(u64::try_from(alpha_num).unwrap_or(0));
        if left <= right {
            rejected.insert(outcome_id);
        } else {
            break;
        }
    }
    rejected
}

#[cfg(test)]
mod tests {
    use super::{OutcomeFamily, SeedPair, StatError, StatisticalPlan, aggregate};
    use std::collections::BTreeMap;

    fn diagnostic_plan() -> StatisticalPlan {
        StatisticalPlan {
            preregistered_outcomes: vec!["latency".to_owned(), "reward".to_owned()],
            declared_pairs: vec![
                "p1".to_owned(),
                "p2".to_owned(),
                "p3".to_owned(),
                "p4".to_owned(),
            ],
            declared_exclusions: vec!["p_skip".to_owned()],
            bootstrap_resamples: 20,
            bootstrap_seed: 1,
            alpha_numerator: 5,
            alpha_denominator: 100,
        }
    }

    fn pair(
        id: &str,
        left: Option<i64>,
        right: Option<i64>,
        failed_a: bool,
        failed_b: bool,
    ) -> SeedPair {
        SeedPair {
            pair_id: id.to_owned(),
            outcome_a: left,
            outcome_b: right,
            failed_a,
            failed_b,
        }
    }

    fn diagnostic_families() -> Vec<OutcomeFamily> {
        vec![
            OutcomeFamily {
                outcome_id: "latency".to_owned(),
                pairs: vec![
                    pair("p1", Some(3), Some(4), false, false),
                    pair("p2", Some(2), Some(5), false, false),
                    pair("p3", Some(1), Some(6), false, false),
                    pair("p4", None, Some(2), true, false),
                    pair("p_skip", Some(0), Some(50), false, false),
                ],
            },
            OutcomeFamily {
                outcome_id: "reward".to_owned(),
                pairs: vec![
                    pair("p1", Some(10), Some(6), false, false),
                    pair("p2", Some(8), Some(7), false, false),
                    pair("p3", Some(12), Some(5), false, false),
                    pair("p4", Some(9), None, false, true),
                    pair("p_skip", Some(100), Some(0), false, false),
                ],
            },
        ]
    }

    #[test]
    fn undeclared_exclusion_fails_closed() {
        let mut families = diagnostic_families();
        families[0]
            .pairs
            .push(pair("sneak", Some(1), Some(0), false, false));
        assert_eq!(
            aggregate(&diagnostic_plan(), &families).err(),
            Some(StatError::UndeclaredExclusion)
        );
    }

    #[test]
    fn undeclared_metric_fails_closed() {
        let mut families = diagnostic_families();
        families.push(OutcomeFamily {
            outcome_id: "posthoc".to_owned(),
            pairs: vec![pair("p1", Some(1), Some(0), false, false)],
        });
        assert_eq!(
            aggregate(&diagnostic_plan(), &families).err(),
            Some(StatError::UndeclaredMetric)
        );
    }

    #[test]
    fn committed_corpus_matches_independent_python_reference() {
        let expected: serde_json::Value = serde_json::from_str(include_str!(
            "../../../fixtures/statistical-aggregation/v1/expected-outcomes.json"
        ))
        .expect("fixture");
        let table = aggregate(&diagnostic_plan(), &diagnostic_families()).expect("stats");
        let rows = table
            .rows
            .iter()
            .map(|row| {
                (
                    row.outcome_id.clone(),
                    serde_json::to_value(row).expect("row"),
                )
            })
            .collect::<BTreeMap<_, _>>();
        assert_eq!(
            serde_json::json!({"schema":"bonsai.statistical-outcomes/v1","rows":rows}),
            expected
        );
    }
}
